### PR TITLE
Fix: Never cache watermarked files, add reload() method

### PR DESCRIPTION
**Pull request by tonyjin**
- Never cache watermarked files
- Separate loading and re-loading (when cache is stale or file is newly watermarked) when loading from server
- Remove superfluous caching before loading from cache
- Add new public method to reload current preview and replace legacy methods of reloading preview with new method

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @tonyjin has signed the CLA. Thanks for the pull request!

**Comment by tonyjin** [comment]
Looking for reviews before I add tests. When I dug into this issue, two things came up:
- We cache and then immediately uncache and reload watermarked files (this was causing the bug with double images when you open a preview of a file before a watermarked image and then navigate to it and rotate)
- We don't distinguish between when we need to load a viewer for the first time (file isn't cached) and when we need to reload (cache is stale or file is newly watermarked)

This patch aims to fix both of those issues and also exposes a new public `reload()` method that can be used to reload the current preview, optionally refreshing file info from the server. Because I'd like to keep some compatibility with our existing `skipServerUpdate` option, `reload()` with no params respects `skipServerUpdate`.

File: src/lib/viewers/BaseViewer.js [hunk 1]
```diff
@@ -175,7 +175,9 @@ class BaseViewer extends EventEmitter {
         }
 
         const iconWrapperEl = container.querySelector(SELECTOR_BOX_PREVIEW_ICON);
-        iconWrapperEl.innerHTML = this.fileLoadingIcon || getIconFromName('FILE_DEFAULT');
+        if (iconWrapperEl) {
+            iconWrapperEl.innerHTML = this.fileLoadingIcon || getIconFromName('FILE_DEFAULT');
```
Comment (JustinHoldstock) [inline src/lib/viewers/BaseViewer.js]: Do we have a constant anywhere for `'FILE_DEFAULT'`?

File: src/lib/Preview.js [hunk 2]
```diff
@@ -779,35 +834,33 @@ class Preview extends EventEmitter {
...
             this.logger.setFile(file);
 
             // Keep reference to previously cached file version
             const cachedFile = this.cache.get(file.id);
 
             // Explicitly uncache watermarked files, otherwise update cache
-            const isWatermarked = file.watermark_info && file.watermark_info.is_watermarked;
-            if (isWatermarked) {
+            const isFileWatermarked = isWatermarked(file);
+            if (isFileWatermarked) {
```
Comment (JustinHoldstock) [inline src/lib/Preview.js]: Unless you're using `isFileWatermarked` elsewhere, could you just put the `isWatermarked()` call into the if?

File: src/lib/Preview.js [hunk 3]
```diff
@@ -213,6 +222,44 @@ class Preview extends EventEmitter {
...
+     *   - skipServerUpdate is true (either passed in or defined in preview options) AND cached file is not valid
+     *   - skipServerUpdate is false AND there is no cached file ID
+     *
+     * @public
+     * @param {boolean} skipServerUpdate - Whether or not to update file info from server
+     * @return {void}
+     */
+    reload(skipServerUpdate) {
+        // If not passed in, default to Preview option for skipping server update
+        if (typeof skipServerUpdate === 'undefined') {
```
Comment (JustinHoldstock) [inline src/lib/Preview.js]: If someone passes in `null`, `this.options.skipServerUpdate` won't be respected

Comment (jeremypress) [reply src/lib/Preview.js]: We could check if it's not a boolean

File: src/lib/Preview.js (hunk 2 above)
Comment (tonyjin) [reply src/lib/Preview.js]: Used a few lines below

File: src/lib/Preview.js (hunk 3 above)
Comment (tonyjin) [reply src/lib/Preview.js]: I think that's fine, you'd have to explicitly pass in `null` and in that case, it evaluates to falsy.

File: src/lib/viewers/BaseViewer.js (hunk 1 above)
Comment (tonyjin) [reply src/lib/viewers/BaseViewer.js]: Nope, I can add it. We're not super consistent about what's defined as a constant vs what's not.

**Comment by tonyjin** [comment]
@jeremypress `reload()` has the benefit of not requiring the user to pass in file ID, authentication, or any preview options.

**Pull request merged**